src/lib/homestore/heap_chunk_selector.cpp

```diff
@@ -0,0 +1,127 @@
+    uint32_t pdevID = 0;
+    auto& pdevIdHint = hint.pdev_id_hint;
+    if (!pdevIdHint.has_value()) {
+        // this is the first shard of this pg, select a pdev with the most available blocks for it
+        auto&& it =
+            std::max_element(m_per_dev_heap.begin(), m_per_dev_heap.end(),
+                             [](const std::pair< const uint32_t, std::shared_ptr< PerDevHeap > >& lhs,
+                                const std::pair< const uint32_t, std::shared_ptr< PerDevHeap > >& rhs) {
+                                 return lhs.second->available_blk_count.load() < rhs.second->available_blk_count.load();
+                             });
+        if (it == m_per_dev_heap.end()) {
+            LOGINFO("No pdev found for new pg");
+            return nullptr;
+        }
+        pdevID = it->first;
+    } else {
+        pdevID = pdevIdHint.value();
+    }
+
+    auto it = m_per_dev_heap.find(pdevID);
+    if (it == m_per_dev_heap.end()) {
+        LOGINFO("No pdev found for pdev {}", pdevID);
+        return nullptr;
+    }
+
+    const auto& vchunk = [it = std::move(it), pdevID]() {
+        auto& heapLock = it->second->mtx;
+        auto& heap = it->second->m_heap;
+        std::lock_guard< std::mutex > l(heapLock);
+        if (heap.empty()) return VChunk(nullptr);
+        VChunk vchunk = heap.top();
+        heap.pop();
+        return vchunk;
+    }();
+
+    if (vchunk.get_internal_chunk()) {
+        auto& avalableBlkCounter = it->second->available_blk_count;
+        avalableBlkCounter.fetch_sub(vchunk.available_blks());
+    } else {
+        LOGINFO("No pdev found for pdev {}", pdevID);
+    }
+
+    return vchunk.get_internal_chunk();
+}
+
+void HeapChunkSelector::foreach_chunks(std::function< void(csharedChunk&) >&& cb) {
+    // we should call `cb` on all the chunks, selected or not
+    std::for_each(std::execution::par_unseq, m_chunks.begin(), m_chunks.end(),
+                  [cb = std::move(cb)](auto& p) { cb(p.second); });
+}
+
+void HeapChunkSelector::release_chunk(const uint16_t chunkID) {
+    const auto& it = m_chunks.find(chunkID);
+    if (it == m_chunks.end()) {
+        LOGWARN("No chunk found for ChunkID {}", chunkID);
+    } else {
+        add_chunk(it->second);
+    }
+}
+
+void HeapChunkSelector::mark_chunk_selected(const uint16_t chunkID) {
+    const auto& it = m_chunks.find(chunkID);
+    if (it == m_chunks.end()) {
+        LOGWARN("No chunk found for ChunkID {}", chunkID);
+        return;
+    }
+    uint32_t pdevID = VChunk(m_chunks[chunkID]).get_pdev_id();
+    auto& heapLock = m_per_dev_heap[pdevID]->mtx;
+    auto& heap = m_per_dev_heap[pdevID]->m_heap;
+    std::vector< VChunk > temp;
+    std::lock_guard< std::mutex > l(heapLock);
+    for (; !heap.empty();) {
+        VChunk vchunk = heap.top();
+        heap.pop();
+        if (vchunk.get_chunk_id() == chunkID) break;
+        temp.emplace_back(vchunk);
+    }
+    for (auto& vchunk : temp) {
```

Comment:
it is also required to reduce the pdev available blk num, right?

Reply:
good catch , thanks , will fix it